dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/DataExtensionProvider.java

```diff
@@ -242,21 +242,11 @@ public void applicationStarting(ApplicationInfo appInfo) throws StateChangeExcep
     public void applicationStarted(ApplicationInfo appInfo) throws StateChangeException {
         Collection<FutureEMBuilder> futures = futureEMBuilders.remove(appInfo.getName());
         if (futures != null) {
-            boolean beforeCheckpoint = !CheckpointPhase.getPhase().restored();
             for (FutureEMBuilder futureEMBuilder : futures) {
-                if (beforeCheckpoint)
-                    try {
-                        // Run the task in the foreground if before a checkpoint.
-                        // This is necessary to ensure this task completes before the checkpoint.
-                        // Application startup performance is not as important before checkpoint
-                        // and this ensures we don't do this work on restore side which will make
-                        // restore faster.
-                        futureEMBuilder.complete(futureEMBuilder.createEMBuilder());
-                    } catch (Throwable x) {
-                        futureEMBuilder.completeExceptionally(x);
-                    }
-                else
-                    futureEMBuilder.completeAsync(futureEMBuilder::createEMBuilder, executor);
+                // This delays createEMBuilder until restore.
+                // While this works by avoiding all connections to the data source, it does make restore much slower.
+                // TODO figure out how to do more work on restore without having to make a connection to the data source
+                CheckpointPhase.onRestore(() -> futureEMBuilder.completeAsync(futureEMBuilder::createEMBuilder, executor));
```

Comment:
This appears to be okay because CheckpointPhase.onRestore is documented as running the operation immediately if not using a checkpoint/restore.  This is very confusing and when we remove the TODO in a future PR we should add some explanation that it will do this so future maintainers can understand this code.

Reply:
I'm open to another name for `onRestore` if it can make things more clear.  I didn't really like `runNowOrOnRestore`.

Reply:
I don't have very good ideas either - `CheckpointPhase.onRestoreElseRun(...)`.  We will either need to find a name that makes it clear that it runs the code even if not using checkpoint/restore, or otherwise ensure we include comments where it is used. The existing name is too misleading and makes it look like the supplied code is only for the restore path.